### PR TITLE
T-51 Changes to asset controller

### PR DESCRIPTION
**Pull request by DashaDrahun**
AddConcreteAsset moved to AssetService. public ActionResult Details(int id) and Delete(int id)  done in assetService method which returns viewmodel. GetAssetPropertiesTypes renamed to SelectAssetPropertyTypes. private readonly AddingAssetHelper assetHelper renamed to AssetHelper. Added tests

### CODE REVIEW

File: PortfolioKeeper.Tests/BusinessLogic/Mappers/AssetTypeMapperTests.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                ParrentTypeId = null
+            };
+            underTest = new AssetTypeMapper(mapper.Object);
+        }
+
+        [Test]
+        public void MapType_When_AssetTypeCollection_Is_Not_Null_Should_Return_CollectionAssetTypeViewModel()
+        {
+            //arrange
+            var propertyTypes = new List<AssetPropertyType>
```
Comment (AntonBessarabov) [inline PortfolioKeeper.Tests/BusinessLogic/Mappers/AssetTypeMapperTests.cs]: `assetType.PropertyTypes = new ...`

File: PortfolioKeeper.Tests/BusinessLogic/Mappers/PortfolioMapperTests.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        private PortfolioMapper underTest;
+        private Portfolio portfolio;
+        private Mock<IAssetMapper> mapper;
+
+        [SetUp]
+        public void SetUp()
+        {
+            mapper = new Mock<IAssetMapper>();
+            mapper.Setup(m => m.Map(It.IsAny<Asset>())).Returns(It.IsAny<AssetViewModel>());
```
Comment (AntonBessarabov) [inline PortfolioKeeper.Tests/BusinessLogic/Mappers/PortfolioMapperTests.cs]: Unnecessary setup

File: PortfolioKeeper.Tests/BusinessLogic/Mappers/AssetTypeMapperTests.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        private Mock<IAssetPropertyMapper> mapper;
+        private AssetType assetType;
+        private IAssetTypeMapper underTest;
+
+        [SetUp]
+        public void SetUp()
+        {
+            mapper = new Mock<IAssetPropertyMapper>();
+            mapper.Setup(m => m.MapType(It.IsAny<AssetPropertyType>()))
+                .Returns(It.IsAny<AssetPropertyTypeViewModel>());
```
Comment (AntonBessarabov) [inline PortfolioKeeper.Tests/BusinessLogic/Mappers/AssetTypeMapperTests.cs]: Unnecessary setup

File: PortfolioKeeper.Tests/BusinessLogic/Mappers/PortfolioMapperTests.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                AppUser = new AppUser()
+            };
+            underTest = new PortfolioMapper(mapper.Object);
+        }
+
+        [Test]
+        public void Map_When_Portfolio_Is_Not_Null_Should_Return_PortfolioViewModel()
+        {
+            //arrange
+            var assets = new List<Asset>
```
Comment (AntonBessarabov) [inline PortfolioKeeper.Tests/BusinessLogic/Mappers/PortfolioMapperTests.cs]: `portfolio.Assets = new...`

File: PortfolioKeeper.Tests/BusinessLogic/Mappers/PortfolioMapperTests.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                    Id = 1
+                },
+                new Asset
+                {
+                    PortfolioId = 1,
+                    Name = "secondAsset",
+                    Id = 2
+                }
+            };
+            portfolio.Assets = assets;
```
Comment (AntonBessarabov) [inline PortfolioKeeper.Tests/BusinessLogic/Mappers/PortfolioMapperTests.cs]: Delete, please

File: PortfolioKeeper.Tests/BusinessLogic/Mappers/AssetTypeMapperTests.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                {
+                    AssetTypeId = 1,
+                    DisplayName = "SecondProp",
+                    DisplayOrder = 2,
+                    Id = 2,
+                    Name = "SecondProp"
+                }
+            };
+
+            assetType.PropertyTypes = propertyTypes;
```
Comment (AntonBessarabov) [inline PortfolioKeeper.Tests/BusinessLogic/Mappers/AssetTypeMapperTests.cs]: Delete, please

File: PortfolioKeeper.Tests/BusinessLogic/Mappers/PropertyTypeValueMapperTests.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            };
+
+            // act
+            var result = underTest.MapValue(propertyValue);
+
+            // assert
+            result.AssetId.Should().Be(propertyValue.AssetId);
+            result.Value.Should()
+                .Be(
+                    $"<value type=\"{propertyValue.ValueType}\">{propertyValue.Value}</value>");
```
Comment (AntonBessarabov) [inline PortfolioKeeper.Tests/BusinessLogic/Mappers/PropertyTypeValueMapperTests.cs]: `               .Be($"<value type=\"{propertyValue.ValueType}\">{propertyValue.Value}</value>");`

File: PortfolioKeeper.Tests/BusinessLogic/Services/PortfolioServiceTests.cs [hunk 8]
```diff
@@ -23,16 +28,17 @@ public void SetUp()
         {
             uof = new Mock<IUnitOfWork>();
             mapper = new Mock<IMapper>();
+            portfolioMapper = new Mock<IPortfolioMapper>();
             portfolioRepository = new Mock<IRepository<Portfolio, int>>();
             uof.Setup(x => x.GetRepository<Portfolio, int>()).Returns(portfolioRepository.Object);
```
Comment (AntonBessarabov) [inline PortfolioKeeper.Tests/BusinessLogic/Services/PortfolioServiceTests.cs]: Separate by empty lines, please

File: PortfolioKeeper/BusinessLogic/Mappers/AssetMapper.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+                Name = asset.Name,
+                PortfolioId = asset.PortfolioId
+            };
+            var lastAssetValuation = asset.Valuations?.LastOrDefault();
+
+            if (lastAssetValuation == null)
+                return assetViewModel;
+
+            assetViewModel.Value = lastAssetValuation.Value;
+            assetViewModel.CurrencyName = lastAssetValuation.Currency?.Name;
```
Comment (AntonBessarabov) [inline PortfolioKeeper/BusinessLogic/Mappers/AssetMapper.cs]: CurrencyName could be null?

File: PortfolioKeeper/BusinessLogic/Mappers/PortfolioMapper.cs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+            if (portfolio.Assets != null)
+                portfolioViewModel.Assets = portfolio.Assets.Select(x => assetMapper.Map(x)).ToList();
+
+            var lastPortfolioValuation = portfolio.Valuations?.LastOrDefault();
+
+            if (lastPortfolioValuation == null)
+                return portfolioViewModel;
+
+            portfolioViewModel.Value = lastPortfolioValuation.Value;
+            portfolioViewModel.CurrencyName = lastPortfolioValuation.Currency?.Name;
```
Comment (AntonBessarabov) [inline PortfolioKeeper/BusinessLogic/Mappers/PortfolioMapper.cs]: CurrencyName could be null?

File: PortfolioKeeper/BusinessLogic/Services/AssetService.cs [hunk 11]
```diff
@@ -27,21 +30,32 @@ public class AssetService : IAssetService
         private readonly IRepository<AssetValuation, int> assetValuationRepository;
         private readonly IRepository<AssetPropertyType, int> assetPropertyTypeRepository;
         private readonly IRepository<Portfolio, int> portfolioRepository;
+        private readonly IAssetHelper assetHelper;
 
 
         public AssetService(IUnitOfWork unitOfWork,
```
Comment (AntonBessarabov) [inline PortfolioKeeper/BusinessLogic/Services/AssetService.cs]: Move `IUnitOfWork unitOfWork,` to a new line, please

File: PortfolioKeeper/BusinessLogic/Services/AssetService.cs [hunk 12]
```diff
@@ -243,6 +247,37 @@ public IEnumerable<AssetPropertyValueDto> GetAssetPropertyValues(int assetId, st
...
+
+            return model;
+        }
+
+        public AssetDeleteViewModel GetAssetForDelete(int id)
+        {
+            var dto = GetAsset(id);
+
+            if (dto == null)
+                throw new NullReferenceException("Asset is not found");
```
Comment (AntonBessarabov) [inline PortfolioKeeper/BusinessLogic/Services/AssetService.cs]: I'm not sure that `NullReferenceException` is good choice here

File: PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs [hunk 13]
```diff
@@ -14,22 +15,27 @@ public class PortfolioService : IPortfolioService
         private static readonly Logger logger = LogManager.GetCurrentClassLogger();
         private readonly IUnitOfWork unitOfWork;
         private readonly IMapper mapper;
+        private readonly IPortfolioMapper portfolioMapper;
         private readonly IRepository<Portfolio, int> portfolioRepository;
-        
+
         public PortfolioService(IUnitOfWork unitOfWork,
```
Comment (AntonBessarabov) [inline PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs]: Move `IUnitOfWork unitOfWork,` to a new line, please.

File: PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs [hunk 14]
```diff
@@ -47,57 +53,14 @@ public void Create(PortfolioDto portfolioDto)
...
 
-        public PortfolioDto Overview(string userId)
+        public PortfolioViewModel Overview(string userId)
         {
-            var portfolios = portfolioRepository.Get(e => e.AppUser.Id == userId).ToList();
+            var portfolio = portfolioRepository.Get(e => e.AppUser.Id == userId).FirstOrDefault();
 
-            if (portfolios == null || !portfolios.Any())
-            {
+            if (portfolio == null)
```
Comment (AntonBessarabov) [inline PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs]: Optional: you can use `??` operator here

File: PortfolioKeeper/Controllers/AssetController.cs [hunk 15]
```diff
@@ -18,41 +16,40 @@ namespace PortfolioKeeper.Controllers
...
+        private readonly IAssetHelper assetHelper;
         private readonly ISecurityCheckService securityService;
 
         public AssetController(
             IAssetService assetService,
-            AddingAssetHelper assetHelper,
-            IAssetMapper assetMapper,
+            IAssetHelper assetHelper,
+            IAssetTypeMapper assetTypeMapper,
             IMapper mapper,
```
Comment (AntonBessarabov) [inline PortfolioKeeper/Controllers/AssetController.cs]: Delete unnecessary args, please

File: PortfolioKeeper/BusinessLogic/Mappers/AssetMapper.cs (hunk 9 above)
Comment (DashaDrahun) [reply PortfolioKeeper/BusinessLogic/Mappers/AssetMapper.cs]: virtual field Currency can be null

File: PortfolioKeeper/BusinessLogic/Mappers/PortfolioMapper.cs (hunk 10 above)
Comment (DashaDrahun) [reply PortfolioKeeper/BusinessLogic/Mappers/PortfolioMapper.cs]: virtual field Currency can be null

File: PortfolioKeeper/BusinessLogic/Services/AssetService.cs (hunk 12 above)
Comment (DashaDrahun) [reply PortfolioKeeper/BusinessLogic/Services/AssetService.cs]: Later it will be custom type of exception here

**Pull request closed without merging**